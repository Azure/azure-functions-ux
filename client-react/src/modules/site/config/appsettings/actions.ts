import { IAction } from '../../../../models/action';
import { IAppSettingsState } from './reducer';

export const UPDATE_SITE_APP_SETTINGS = 'UPDATE_SITE_APP_SETTINGS';
export const updateCurrentSiteAppSettings = (settings: Partial<IAppSettingsState>): IAction<Partial<IAppSettingsState>> => ({
  payload: settings,
  type: UPDATE_SITE_APP_SETTINGS,
});

export const UPDATE_SITE_APP_SETTINGS_NO_CACHE = 'UPDATE_SITE_APP_SETTINGS_NO_CACHE';
export const updateCurrentSiteAppSettingsNoCache = (settings: Partial<IAppSettingsState>): IAction<Partial<IAppSettingsState>> => ({
  payload: settings,
  type: UPDATE_SITE_APP_SETTINGS,
});
