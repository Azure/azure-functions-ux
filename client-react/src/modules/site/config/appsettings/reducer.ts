import { combineReducers } from 'redux';
import { ActionType } from 'typesafe-actions';

import { ArmObj } from '../../../../models/WebAppModels';
import { metadataReducer } from '../../../ApiReducerHelper';
import { ApiState } from '../../../types';
import * as actions from './actions';
import { APP_SETTINGS_FETCH_SUCCESS, APP_SETTINGS_UPDATE_SUCCESS, AREA_STRING, UPDATE_APP_SETTINGS_FROM_SITE_UPDATE } from './actionTypes';

export type AppSettings = { [key: string]: string };

export type AppSettingsActions = ActionType<typeof actions>;
export type AppSettingsState = ApiState<ArmObj<AppSettings>>;

export const InitialState = {
  data: {
    id: '',
    properties: {},
    name: '',
    location: '',
    kind: '',
  },
};

export default combineReducers<AppSettingsState, AppSettingsActions>({
  metadata: metadataReducer(AREA_STRING),
  data: (state = InitialState.data, action) => {
    switch (action.type) {
      case APP_SETTINGS_FETCH_SUCCESS:
        return action.appSettings;
      case APP_SETTINGS_UPDATE_SUCCESS:
        return action.appSettings;
      case UPDATE_APP_SETTINGS_FROM_SITE_UPDATE:
        return { ...state, properties: action.appSettings ? action.appSettings : state.properties };
      default:
        return state;
    }
  },
});
