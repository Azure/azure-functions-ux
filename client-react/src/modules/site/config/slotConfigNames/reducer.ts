import { IAction } from '../../../../models/action';
import { DEFAULT_KEY, generateCacheTTL } from 'redux-cache';
import { UPDATE_SITE_SLOT_CONFIG_NAMES, UPDATE_SITE_SLOT_CONFIG_NAMES_NO_CACHE } from './actions';
import { SlotConfigNames } from '../../../../models/WebAppModels';

export interface ISlotConfigNamesState {
  loading: boolean;
  saving: boolean;
  resourceId: string;
  slotConfigNames: SlotConfigNames;
}
export const InitialState: ISlotConfigNamesState = {
  [DEFAULT_KEY]: null,
  loading: false,
  resourceId: '',
  saving: false,
  slotConfigNames: {
    appSettingNames: [],
    connectionStringNames: [],
    azureStorageConfigNames: [],
  },
};

const slotConfigNames = (state = InitialState, action: IAction<any>) => {
  switch (action.type) {
    case UPDATE_SITE_SLOT_CONFIG_NAMES:
      return {
        ...state,
        [DEFAULT_KEY]: generateCacheTTL(60000),
        ...action.payload,
      };
    case UPDATE_SITE_SLOT_CONFIG_NAMES_NO_CACHE:
      return {
        ...state,
        ...action.payload,
      };
    default:
      return state;
  }
};

export default slotConfigNames;
