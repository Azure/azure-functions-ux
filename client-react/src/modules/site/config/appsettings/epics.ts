import { combineEpics, Epic } from 'redux-observable';
import { from, of } from 'rxjs';
import { catchError, filter, map, switchMap } from 'rxjs/operators';
import { isActionOf } from 'typesafe-actions';

import * as Types from '../../../types';
import {
  fetchAppSettingsFailure,
  fetchAppSettingsRequest,
  fetchAppSettingsSuccess,
  updateAppSettingsFailure,
  updateAppSettingsRequest,
  updateAppSettingsSuccess,
} from './actions';
import { AppSettingsActions } from './reducer';

export const fetchAppSettings: Epic<AppSettingsActions, AppSettingsActions, Types.RootState, Types.Services> = (
  action$,
  store,
  { appSettingsApi }
) =>
  action$.pipe(
    filter(isActionOf(fetchAppSettingsRequest)),
    switchMap(() =>
      from(appSettingsApi.fetchAppSettings(store.value)).pipe(
        map(fetchAppSettingsSuccess),
        catchError(err => of(fetchAppSettingsFailure(err)))
      )
    )
  );

export const updateAppSettings: Epic<AppSettingsActions, AppSettingsActions, Types.RootState, Types.Services> = (
  action$,
  store,
  { appSettingsApi }
) =>
  action$.pipe(
    filter(isActionOf(updateAppSettingsRequest)),
    switchMap(action =>
      from(appSettingsApi.updateAppSettings(store.value, action.appSettings)).pipe(
        map(updateAppSettingsSuccess),
        catchError(err => of(updateAppSettingsFailure(err)))
      )
    )
  );
export default combineEpics(fetchAppSettings, updateAppSettings);
