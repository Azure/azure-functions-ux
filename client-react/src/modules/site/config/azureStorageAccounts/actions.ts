import { createStandardAction } from 'typesafe-actions';

import { ArmObj, Site } from '../../../../models/WebAppModels';
import {
  AZURE_STORAGE_MOUNT_FETCH_FAILURE,
  AZURE_STORAGE_MOUNT_FETCH_REQUEST,
  AZURE_STORAGE_MOUNT_FETCH_SUCCESS,
  AZURE_STORAGE_MOUNT_UPDATE_FAILURE,
  AZURE_STORAGE_MOUNT_UPDATE_REQUEST,
  AZURE_STORAGE_MOUNT_UPDATE_SUCCESS,
  UPDATE_AZURE_STORAGE_MOUNT_FROM_SITE_UPDATE,
} from './actionTypes';
import { ArmAzureStorageMount } from './reducer';

export const fetchAzureStorageMountRequest = createStandardAction(AZURE_STORAGE_MOUNT_FETCH_REQUEST)();

export const fetchAzureStorageMountSuccess = createStandardAction(AZURE_STORAGE_MOUNT_FETCH_SUCCESS).map(
  (azureMount: ArmObj<ArmAzureStorageMount>) => ({
    azureMount,
  })
);
export const fetchAzureStorageMountFailure = createStandardAction(AZURE_STORAGE_MOUNT_FETCH_FAILURE).map((error: Error) => ({
  error,
}));

export const updateAzureStorageMountRequest = createStandardAction(AZURE_STORAGE_MOUNT_UPDATE_REQUEST).map(
  (azureMount: ArmObj<ArmAzureStorageMount>) => ({
    azureMount,
  })
);

export const updateAzureStorageMountSuccess = createStandardAction(AZURE_STORAGE_MOUNT_UPDATE_SUCCESS).map(
  (azureMount: ArmObj<ArmAzureStorageMount>) => ({
    azureMount,
  })
);

export const updateAzureStorageMountFailure = createStandardAction(AZURE_STORAGE_MOUNT_UPDATE_FAILURE).map((error: Error) => ({
  error,
}));

export const updateAzureStorageMountFromSiteUpdate = createStandardAction(UPDATE_AZURE_STORAGE_MOUNT_FROM_SITE_UPDATE).map(
  (site: ArmObj<Site>) => {
    const azureStorageMount = !!site.properties && !!site.properties.siteConfig && site.properties.siteConfig.azureStorageAccounts;
    return { azureStorageMount };
  }
);
