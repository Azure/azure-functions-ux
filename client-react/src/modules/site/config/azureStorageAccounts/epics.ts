import { combineEpics, Epic } from 'redux-observable';
import { from, of } from 'rxjs';
import { catchError, filter, map, switchMap } from 'rxjs/operators';
import { isActionOf } from 'typesafe-actions';

import * as Types from '../../../types';
import {
  fetchAzureStorageMountRequest,
  fetchAzureStorageMountSuccess,
  fetchAzureStorageMountFailure,
  updateAzureStorageMountRequest,
  updateAzureStorageMountSuccess,
  updateAzureStorageMountFailure,
} from './actions';
import { AzureStorageMountActions } from './reducer';

export const fetchAzureStorageMount: Epic<AzureStorageMountActions, AzureStorageMountActions, Types.RootState, Types.Services> = (
  action$,
  store,
  { azureMountApi }
) =>
  action$.pipe(
    filter(isActionOf(fetchAzureStorageMountRequest)),
    switchMap(() =>
      from(azureMountApi.fetchAzureStorageMount(store.value)).pipe(
        map(fetchAzureStorageMountSuccess),
        catchError(err => of(fetchAzureStorageMountFailure(err)))
      )
    )
  );

export const updateAzureStorageMount: Epic<AzureStorageMountActions, AzureStorageMountActions, Types.RootState, Types.Services> = (
  action$,
  store,
  { azureMountApi }
) =>
  action$.pipe(
    filter(isActionOf(updateAzureStorageMountRequest)),
    switchMap(action =>
      from(azureMountApi.updateAzureStorageMount(store.value, action.azureMount)).pipe(
        map(updateAzureStorageMountSuccess),
        catchError(err => of(updateAzureStorageMountFailure(err)))
      )
    )
  );
export default combineEpics(fetchAzureStorageMount, updateAzureStorageMount);
