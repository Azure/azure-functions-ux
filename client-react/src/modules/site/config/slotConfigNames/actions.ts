import { IAction } from '../../../../models/action';
import { ISlotConfigNamesState } from './reducer';

export const UPDATE_SITE_SLOT_CONFIG_NAMES = 'UPDATE_SITE_SLOT_CONFIG_NAMES';
export const updateCurrentSiteSlotConfigNames = (
  slotConfigNames: Partial<ISlotConfigNamesState>
): IAction<Partial<ISlotConfigNamesState>> => ({
  payload: slotConfigNames,
  type: UPDATE_SITE_SLOT_CONFIG_NAMES,
});

export const UPDATE_SITE_SLOT_CONFIG_NAMES_NO_CACHE = 'UPDATE_SITE_SLOT_CONFIG_NAMES_NO_CACHE';
export const updateCurrentSiteSlotConfigNamesNoCache = (
  slotConfigNames: Partial<ISlotConfigNamesState>
): IAction<Partial<ISlotConfigNamesState>> => ({
  payload: slotConfigNames,
  type: UPDATE_SITE_SLOT_CONFIG_NAMES_NO_CACHE,
});
