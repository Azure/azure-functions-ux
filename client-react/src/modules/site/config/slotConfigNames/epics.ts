import { combineEpics, Epic } from 'redux-observable';
import { from, of } from 'rxjs';
import { catchError, filter, map, switchMap } from 'rxjs/operators';
import { isActionOf } from 'typesafe-actions';

import { RootState, Services } from '../../../types';
import {
  fetchSlotConfigFailure,
  fetchSlotConfigRequest,
  fetchSlotConfigSuccess,
  updateSlotConfigFailure,
  updateSlotConfigRequest,
  updateSlotConfigSuccess,
} from './actions';
import { SlotConfigAction } from './reducer';

export const fetchSlotConfigName: Epic<SlotConfigAction, SlotConfigAction, RootState, Services> = (
  action$,
  store,
  { slotConfigNamesApi }
) =>
  action$.pipe(
    filter(isActionOf(fetchSlotConfigRequest)),
    switchMap(() =>
      from(slotConfigNamesApi.fetchSlotConfig(store.value)).pipe(
        map(fetchSlotConfigSuccess),
        catchError(err => of(fetchSlotConfigFailure(err)))
      )
    )
  );

export const updateSlotConfigName: Epic<SlotConfigAction, SlotConfigAction, RootState, Services> = (
  action$,
  store,
  { slotConfigNamesApi }
) =>
  action$.pipe(
    filter(isActionOf(updateSlotConfigRequest)),
    switchMap(action =>
      from(slotConfigNamesApi.updateSlotConfig(store.value, action.slotConfig)).pipe(
        map(updateSlotConfigSuccess),
        catchError(err => of(updateSlotConfigFailure(err)))
      )
    )
  );
export default combineEpics(fetchSlotConfigName, updateSlotConfigName);
