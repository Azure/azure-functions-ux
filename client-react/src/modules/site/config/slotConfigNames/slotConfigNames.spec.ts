import { ActionsObservable } from 'redux-observable';
import { toArray } from 'rxjs/operators';

import rootReducer from '../../..';
import { IStartupInfo } from '../../../../models/portal-models';
import { ArmObj, SlotConfigNames } from '../../../../models/WebAppModels';
import { getStartupInfoAction } from '../../../portal/actions';
import { RootState, Services } from '../../../types';
import { updateResourceId } from '../../actions';
import {
  fetchSlotConfigFailure,
  fetchSlotConfigRequest,
  fetchSlotConfigSuccess,
  updateSlotConfigFailure,
  updateSlotConfigRequest,
  updateSlotConfigSuccess,
} from './actions';
import {
  SLOT_CONFIG_FETCH_FAILURE,
  SLOT_CONFIG_FETCH_SUCCESS,
  SLOT_CONFIG_UPDATE_FAILURE,
  SLOT_CONFIG_UPDATE_SUCCESS,
} from './actionTypes';
import { fetchSlotConfigName, updateSlotConfigName } from './epics';
import reducer from './reducer';
import api from './slotConfigNamesService';
jest.mock('../../../ArmHelper');
import MakeArmCall from '../../../ArmHelper';

const testResult: ArmObj<SlotConfigNames> = {
  id: '',
  name: '',
  location: '',
  kind: '',
  properties: {
    connectionStringNames: ['testvalue'],
    appSettingNames: ['testvalue'],
    azureStorageConfigNames: ['testvalue'],
  },
};
describe('Slot Config Names Store Epics', () => {
  const successDeps = {
    slotConfigNamesApi: {
      fetchSlotConfig: async (state: RootState): Promise<ArmObj<SlotConfigNames>> => {
        return testResult;
      },
      updateSlotConfig: async (state: RootState, newConfigName: ArmObj<SlotConfigNames>): Promise<ArmObj<SlotConfigNames>> => {
        return testResult;
      },
    },
  } as Services;

  const failDeps = {
    slotConfigNamesApi: {
      fetchSlotConfig: async (state: RootState): Promise<ArmObj<SlotConfigNames>> => {
        throw new Error('failuremessage');
      },
      updateSlotConfig: async (state: RootState, newConfigName: ArmObj<SlotConfigNames>): Promise<ArmObj<SlotConfigNames>> => {
        throw new Error('failuremessage');
      },
    },
  } as Services;
  it('Sends Success Action with object on Successful Fetch', async () => {
    let action$ = ActionsObservable.of(fetchSlotConfigRequest());
    const output$ = fetchSlotConfigName(action$, {} as any, successDeps);
    const actions = await output$.pipe(toArray()).toPromise();
    expect(actions.length).toBe(1);
    const action = actions[0];
    expect(action.type).toBe(SLOT_CONFIG_FETCH_SUCCESS);
    if (action.type === SLOT_CONFIG_FETCH_SUCCESS) {
      expect(action.slotConfig.properties.appSettingNames![0]).toBe('testvalue');
    }
  });

  it('Sends Success Action with object on Successful Update', async () => {
    let action$ = ActionsObservable.of(updateSlotConfigRequest(testResult));
    const output$ = updateSlotConfigName(action$, {} as any, successDeps);
    const actions = await output$.pipe(toArray()).toPromise();
    expect(actions.length).toBe(1);
    const action = actions[0];
    expect(action.type).toBe(SLOT_CONFIG_UPDATE_SUCCESS);
    if (action.type === SLOT_CONFIG_UPDATE_SUCCESS) {
      expect(action.slotConfig.properties.appSettingNames![0]).toBe('testvalue');
    }
  });

  it('Sends Error Action with error on failed Fetch', async () => {
    let action$ = ActionsObservable.of(fetchSlotConfigRequest());
    const output$ = fetchSlotConfigName(action$, {} as any, failDeps);
    const actions = await output$.pipe(toArray()).toPromise();
    expect(actions.length).toBe(1);
    const action = actions[0];
    expect(action.type).toBe(SLOT_CONFIG_FETCH_FAILURE);
    if (action.type === SLOT_CONFIG_FETCH_FAILURE) {
      expect(action.error.message).toBe('failuremessage');
    }
  });

  it('Sends Error Action with error on failed Update', async () => {
    let action$ = ActionsObservable.of(updateSlotConfigRequest(testResult));
    const output$ = updateSlotConfigName(action$, {} as any, failDeps);
    const actions = await output$.pipe(toArray()).toPromise();
    expect(actions.length).toBe(1);
    const action = actions[0];
    expect(action.type).toBe(SLOT_CONFIG_UPDATE_FAILURE);
    if (action.type === SLOT_CONFIG_UPDATE_FAILURE) {
      expect(action.error.message).toBe('failuremessage');
    }
  });
});

describe('Slot Config Names Store Reducer', () => {
  const initialState = reducer(undefined, {} as any);
  describe('initial state', () => {
    it('should match a snapshot', () => {
      expect(initialState).toMatchSnapshot();
    });

    it('loading and updating should be false', () => {
      expect(initialState.metadata.loading).toBe(false);
      expect(initialState.metadata.updating).toBe(false);
    });
  });

  describe('Slot Config Names Fetch Stories', () => {
    it('should trigger loading when the fetch is requested', () => {
      const action = fetchSlotConfigRequest();
      const state = reducer(initialState, action);
      expect(state.metadata.loading).toBe(true);
    });

    it('metadata and app settings should update on successful load', () => {
      const action = fetchSlotConfigSuccess(testResult);
      const state = reducer(initialState, action);
      expect(state.metadata.loading).toBe(false);
      expect(state.data.properties.appSettingNames![0]).toBe('testvalue');
    });

    it('error should be reflected on failed load', () => {
      const action = fetchSlotConfigFailure(new Error('testerror'));
      const state = reducer(initialState, action);
      expect(state.metadata.loading).toBe(false);
      expect(state.metadata.fetchError).toBe(true);
      expect(state.metadata.fetchErrorObject.message).toBe('testerror');
    });
  });

  describe('Slot Config Names UPDATE Stories', () => {
    it('should trigger loading when the update is requested', () => {
      const action = updateSlotConfigRequest(testResult);
      const state = reducer(initialState, action);
      expect(state.metadata.updating).toBe(true);
    });

    it('Slot config names should update on successful load', () => {
      const action = updateSlotConfigSuccess(testResult);
      const state = reducer(initialState, action);
      expect(state.metadata.updating).toBe(false);
      expect(state.data.properties.appSettingNames![0]).toBe('testvalue');
    });

    it('error should be reflected on failed load', () => {
      const action = updateSlotConfigFailure(new Error('testerror'));
      const state = reducer(initialState, action);
      expect(state.metadata.updating).toBe(false);
      expect(state.metadata.updateError).toBe(true);
      expect(state.metadata.updateErrorObject.message).toBe('testerror');
    });
  });
});

describe('Slot Config Names Service', () => {
  const initialState = rootReducer(undefined, {} as any);
  let state;
  beforeEach(() => {
    const updateResourceIdAction = updateResourceId('resourceid');
    const updateSUIAction = getStartupInfoAction({
      token: 'testtoken',
      armEndpoint: 'testEndpoint',
    } as IStartupInfo);

    state = rootReducer(rootReducer(initialState, updateResourceIdAction), updateSUIAction);
  });
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('Fetch Api calls api with appropriate info', async () => {
    api.fetchSlotConfig(state);
    expect(MakeArmCall).toHaveBeenCalledWith('testEndpoint', 'testtoken', 'resourceid/config/slotconfignames', 'FetchSlotConfig');
  });

  it('Fetch Api calls api with appropriate info from slots app', async () => {
    const updateResourceIdAction = updateResourceId('resourceid/slots/slot');
    state = rootReducer(state, updateResourceIdAction);
    api.fetchSlotConfig(state);
    expect(MakeArmCall).toHaveBeenCalledWith('testEndpoint', 'testtoken', 'resourceid/config/slotconfignames', 'FetchSlotConfig');
  });

  it('Update Api calls api with appropriate info', async () => {
    api.updateSlotConfig(state, testResult);
    expect(MakeArmCall).toHaveBeenCalledWith(
      'testEndpoint',
      'testtoken',
      'resourceid/config/slotconfignames',
      'UpdateSlotConfig',
      'PUT',
      testResult
    );
  });

  it('Update Api calls api with appropriate info from slot', async () => {
    const updateResourceIdAction = updateResourceId('resourceid/slots/slot');
    state = rootReducer(state, updateResourceIdAction);
    api.updateSlotConfig(state, testResult);
    expect(MakeArmCall).toHaveBeenCalledWith(
      'testEndpoint',
      'testtoken',
      'resourceid/config/slotconfignames',
      'UpdateSlotConfig',
      'PUT',
      testResult
    );
  });
});
