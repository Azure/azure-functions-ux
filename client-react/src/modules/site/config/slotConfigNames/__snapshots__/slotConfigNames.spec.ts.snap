// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slot Config Names Store Reducer initial state should match a snapshot 1`] = `
Object {
  "data": Object {
    "id": "",
    "kind": "",
    "location": "",
    "name": "",
    "properties": Object {
      "appSettingNames": Array [],
      "azureStorageConfigNames": Array [],
      "connectionStringNames": Array [],
    },
  },
  "metadata": Object {
    "fetchError": false,
    "fetchErrorObject": [Error],
    "loading": false,
    "updateError": false,
    "updateErrorObject": [Error],
    "updating": false,
  },
}
`;
